src/AspNet.Security.OAuth.Shopify/ShopifyAuthenticationHandler.cs

```diff
@@ -0,0 +1,218 @@
+
+            return ticket;
+        }
+
+        /// <inheritdoc />
+        protected override string FormatScope()
+        {
+            return string.Join(",", Options.Scope);
+        }
+
+        /// <inheritdoc />
+        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
+        {
+            if (!properties.Items.ContainsKey(ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty))
+            {
+                var msg =
+                    $"Shopify provider AuthenticationProperties must contain {ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty}";
+
+                Logger.LogError(msg);
+                throw new Exception(msg);
+            }
+
+            var shopName = properties.Items[ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty];
+            var uri = string.Format(Options.AuthorizationEndpoint, shopName);
+
+            // Get the permission scope, which can either be set in options or overridden in AuthenticationProperties.
+            var scope = properties.Items.ContainsKey(ShopifyAuthenticationDefaults.ShopScopeAuthenticationProperty) ? 
+                properties.Items[ShopifyAuthenticationDefaults.ShopScopeAuthenticationProperty] : 
+                FormatScope();
+
+            var url = QueryHelpers.AddQueryString(uri, new Dictionary<string, string>
+            {
+                ["client_id"] = Options.ClientId,
+                ["scope"] = scope,
+                ["redirect_uri"] = redirectUri,
+                ["state"] = Options.StateDataFormat.Protect(properties)
+            });
+
+            // If we're requesting a per-user, online only, token, add the grant_options query param.
+            if (properties.Items.ContainsKey(ShopifyAuthenticationDefaults.GrantOptionsAuthenticationProperty))
+            {
+                var grantOptions = properties.Items[ShopifyAuthenticationDefaults.GrantOptionsAuthenticationProperty];
+                if (grantOptions != null &&
+                    grantOptions.Equals(ShopifyAuthenticationDefaults.PerUserAuthenticationPropertyValue))
+                {
+                    url = QueryHelpers.AddQueryString(url, "grant_options[]",
+                        ShopifyAuthenticationDefaults.PerUserAuthenticationPropertyValue);
+                }
+            }
+
+            return url;
+        }
+
+        /// <inheritdoc />
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(
+            [NotNull] string code, 
+            [NotNull] string redirectUri)
+        {
+            string shopDns;
+
+            try
+            {
+                var shop = Context.Request.Query["shop"];
+                var state = Context.Request.Query["state"];
+
+                // Shop name must end with myshopify.com
+                if (!shop.ToString().EndsWith(".myshopify.com"))
```

Comment:
Specify whether this is case sensitive or not explicitly via `StringComparisonOptions.Ordinal...`.